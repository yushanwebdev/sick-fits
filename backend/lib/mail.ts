/* eslint-disable @typescript-eslint/ban-ts-comment */
/* eslint-disable @typescript-eslint/restrict-template-expressions */
import { createTransport, getTestMessageUrl } from 'nodemailer';

const transporter = createTransport({
  // @ts-ignore
  host: process.env.MAIL_HOST,
  port: process.env.MAIL_PORT,
  auth: {
    user: process.env.MAIL_USER,
    pass: process.env.MAIL_PASS,
  },
});

function makeANiceEmail(text: string) {
  return `
        <div style="
            border: 1px;
            padding: 20px;
            font-family: sans-serif;
            line-height: 2;
            font-size: 20px;
        ">
            <h2>Hello There!</h2>
            <p>${text}</p>
            <p>ðŸš€, Yushan</p>
        </div>
        `;
}

export async function sendPasswordResetEmail(
  resetToken: string,
  to: string
): Promise<void> {
  // email the user a token
  const info = await transporter.sendMail({
    to,
    from: 'yushan@example.com',
    subject: 'Your password reset token!',
    html: makeANiceEmail(`Your Password Reset Token is here!
        <a href="${process.env.FRONTEND_URL}/reset?token=${resetToken}" target="_blank">Click here to reset</a>
    `),
  });

  if (process.env.MAIL_USER.includes('ethereal.email')) {
    console.log(`ðŸ’Œ Message Sent! Preview it at ${getTestMessageUrl(info)}`);
  }
}
