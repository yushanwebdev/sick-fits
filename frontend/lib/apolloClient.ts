// TODO: Fix the type issues
import { ApolloClient, from, InMemoryCache } from "@apollo/client";
import { onError } from "@apollo/client/link/error";
import { createUploadLink } from "apollo-upload-client";
import merge from "deepmerge";
import isEqual from "lodash/isEqual";
import { useMemo } from "react";
import { endpoint, prodEndpoint } from "../config";
import paginationField from "./paginationField";

export const APOLLO_STATE_PROP_NAME = "__APOLLO_STATE__";

let apolloClient;

const errorLink = onError(({ graphQLErrors, networkError }) => {
  if (graphQLErrors)
    graphQLErrors.forEach(({ message, locations, path }) =>
      console.log(
        `[GraphQL error]: Message: ${message}, Location: ${JSON.stringify(
          locations
        )}, Path: ${JSON.stringify(path)}`
      )
    );
  if (networkError)
    console.log(`[Network error]: ${JSON.stringify(networkError)}`);
});

const uploadLink = createUploadLink({
  uri: process.env.NODE_ENV === "development" ? endpoint : prodEndpoint, // Server URL (must be absolute)
  fetchOptions: {
    credentials: "include", // To send the Cookie that stored in the Browser
  },
});

function createApolloClient() {
  return new ApolloClient({
    ssrMode: typeof window === "undefined",
    link: from([errorLink, uploadLink]),
    cache: new InMemoryCache({
      typePolicies: {
        Query: {
          fields: {
            // @ts-ignore
            allProducts: paginationField(),
          },
        },
      },
    }),
  });
}

export function initializeApollo(initialState = null) {
  const _apolloClient = apolloClient ?? createApolloClient();

  // If your page has Next.js data fetching methods that use Apollo Client, the initial state
  // gets hydrated here
  if (initialState) {
    // Get existing cache, loaded during client side data fetching
    const existingCache = _apolloClient.extract();

    // Merge the initialState from getStaticProps/getServerSideProps in the existing cache
    const data = merge(existingCache, initialState, {
      // combine arrays using object equality (like in sets)
      arrayMerge: (destinationArray, sourceArray) => [
        ...sourceArray,
        ...destinationArray.filter((d) =>
          sourceArray.every((s) => !isEqual(d, s))
        ),
      ],
    });

    // Restore the cache with the merged data
    _apolloClient.cache.restore(data);
  }
  // For SSG and SSR always create a new Apollo Client
  if (typeof window === "undefined") return _apolloClient;
  // Create the Apollo Client once in the client
  if (!apolloClient) apolloClient = _apolloClient;

  return _apolloClient;
}

export function addApolloState(client, pageProps) {
  try {
    if (pageProps?.props) {
      pageProps.props[APOLLO_STATE_PROP_NAME] = client.cache.extract() ?? null;
    }
  } catch (error) {
    console.log(error);
  }

  return pageProps;
}

export function useApollo(pageProps) {
  const state = pageProps[APOLLO_STATE_PROP_NAME];
  const store = useMemo(() => initializeApollo(state), [state]);
  return store;
}
