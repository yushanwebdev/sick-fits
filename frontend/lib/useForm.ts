import { ChangeEvent, useState } from 'react';

export default function useForm<T = Record<string, string>>(initial: T) {
  // create a state object for our inputs
  const [inputs, setInputs] = useState(initial);

  function handleChange(
    e: ChangeEvent<HTMLInputElement | HTMLTextAreaElement>
  ) {
    const { value, name, type } = e.target;
    let localValue: string | number | File = value;

    if (type === 'number') {
      localValue = parseInt(value);
    }

    if (type === 'file') {
      [localValue] = (e.target as HTMLInputElement).files;
    }

    setInputs({
      // copy the existing state
      ...inputs,
      [name]: localValue,
    });
  }

  function resetForm() {
    setInputs(initial);
  }

  function clearForm() {
    const blankState = Object.entries(inputs).reduce((acc, [key]) => {
      acc[key] = '';
      return acc;
    }, {} as T);

    setInputs(blankState);
  }

  // return the things we want to surface from this custom hook
  return {
    inputs,
    handleChange,
    resetForm,
    clearForm,
  };
}
