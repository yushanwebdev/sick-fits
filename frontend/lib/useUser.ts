/* eslint-disable @typescript-eslint/no-unsafe-assignment */
/* eslint-disable @typescript-eslint/no-unsafe-member-access */
/* eslint-disable @typescript-eslint/no-unsafe-return */
import { gql, useQuery } from '@apollo/client';

export const CURRENT_USER_QUERY = gql`
  query GetAuthUser {
    authenticatedItem {
      ... on User {
        id
        email
        name
        cart {
          id
          quantity
          product {
            id
            price
            name
            description
            photo {
              image {
                publicUrlTransformed
              }
            }
          }
        }
      }
    }
  }
`;

export const useUser = () => {
  const { data, error } = useQuery(CURRENT_USER_QUERY);

  console.log('error', error);

  return data?.authenticatedItem;
};
