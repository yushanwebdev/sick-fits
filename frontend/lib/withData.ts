import {
  ApolloClient,
  InMemoryCache,
  from,
  NormalizedCacheObject,
  HttpOptions,
} from '@apollo/client';
import { onError } from '@apollo/client/link/error';
import { createUploadLink } from 'apollo-upload-client';
import withApollo from 'next-with-apollo';
import { getDataFromTree } from '@apollo/client/react/ssr';
import { endpoint, prodEndpoint } from '../config';

function createClient({
  headers,
  initialState,
}: {
  headers: HttpOptions['headers'];
  initialState: NormalizedCacheObject;
}) {
  const errorLink = onError(({ graphQLErrors, networkError }) => {
    if (graphQLErrors)
      graphQLErrors.forEach(({ message, locations, path }) =>
        console.log(
          `[GraphQL error]: Message: ${message}, Location: ${JSON.stringify(
            locations
          )}, Path: ${JSON.stringify(path)}`
        )
      );

    if (networkError)
      console.log(`[Network error]: ${JSON.stringify(networkError)}`);
  });

  // this uses apollo-link-http under the hood, so all the options here come from that package
  const uploadLink = createUploadLink({
    uri: process.env.NODE_ENV === 'development' ? endpoint : prodEndpoint,
    credentials: 'include',
    // pass the headers along from this request. This enables SSR with logged in state
    // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
    headers,
  });

  // If you provide a link chain to ApolloClient, you
  // don't provide the `uri` option.
  return new ApolloClient({
    // The `from` function combines an array of individual links
    // into a link chain
    link: from([errorLink, uploadLink]),
    cache: new InMemoryCache().restore(initialState || {}),
  });
}

export default withApollo(createClient, { getDataFromTree });
