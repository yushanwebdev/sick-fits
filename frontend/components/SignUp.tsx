/* eslint-disable @typescript-eslint/no-unsafe-assignment */
/* eslint-disable @typescript-eslint/no-unsafe-member-access */
/* eslint-disable @typescript-eslint/no-unsafe-argument */
import { gql, useMutation } from "@apollo/client";
import useForm from "../lib/useForm";
import Form from "./styles/Form";
import DisplayError from "./DisplayError";

export const SIGN_UP_MUTATION = gql`
  mutation CreateUser($name: String!, $email: String!, $password: String!) {
    createUser(data: { name: $name, email: $email, password: $password }) {
      id
      name
      email
    }
  }
`;

export default function SignUp() {
  const { inputs, handleChange, resetForm } = useForm({
    name: "",
    email: "",
    password: "",
  });
  const [signup, { data, error }] = useMutation(SIGN_UP_MUTATION, {
    variables: inputs,
  });

  const handleSubmit = async (e: React.FormEvent<HTMLFormElement>) => {
    e.preventDefault(); // stop the form from submitting
    // Send the email and password to the graphqlAPI
    await signup().catch(console.error);
    resetForm();
  };

  return (
    <Form method="POST" onSubmit={handleSubmit}>
      <h2>Sign Up For an Account</h2>
      <DisplayError error={error} />
      {data?.createUser ? (
        <p>
          Signed Up with {data.createUser.email} - Please Go Ahead and Sign in!
        </p>
      ) : (
        <fieldset>
          <label htmlFor="name">
            Name
            <input
              type="text"
              name="name"
              placeholder="Your Name"
              autoComplete="name"
              value={inputs.name}
              onChange={handleChange}
            />
          </label>
          <label htmlFor="email">
            Email
            <input
              type="email"
              name="email"
              placeholder="Your Email Address"
              autoComplete="email"
              value={inputs.email}
              onChange={handleChange}
            />
          </label>
          <label htmlFor="password">
            Password
            <input
              type="password"
              name="password"
              placeholder="Password"
              autoComplete="password"
              value={inputs.password}
              onChange={handleChange}
            />
          </label>
          <button type="submit">Sign Up!</button>
        </fieldset>
      )}
    </Form>
  );
}
