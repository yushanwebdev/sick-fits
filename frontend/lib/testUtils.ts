import casual from "casual";
import { PAGINATION_QUERY } from "../components/Pagination";

// seed it so we get consistent results
casual.seed(777);

const fakeItem = () => ({
  __typename: "Item",
  id: "abc123",
  price: 5000,
  user: null,
  photo: {
    __typename: "Photo",
    id: "abc123",
    altText: "dogs are best",
    image: {
      __typename: "Image",
      publicUrlTransformed: "dog.jpg",
    },
  },
  name: "dogs are best",
  description: "dogs",
});

const fakeUser = (overrides?: Object) => ({
  __typename: "User",
  id: "4234",
  name: casual.name,
  email: casual.email,
  permissions: ["ADMIN"],
  orders: [],
  cart: [],
  role: {
    canManageCart: true,
    canManageUsers: true,
    canManageProducts: true,
  },
  ...overrides,
});

const fakeOrderItem = () => ({
  __typename: "OrderItem",
  id: casual.uuid,
  image: {
    image: `${casual.word}.jpg`,
  },
  name: casual.words(),
  price: 4234,
  quantity: 1,
  description: casual.words(),
});

const fakeOrder = () => ({
  __typename: "Order",
  id: "ord123",
  charge: "ch_123",
  total: 40000,
  items: [fakeOrderItem(), fakeOrderItem()],
  createdAt: "2022-12-11T20:16:13.797Z",
  user: fakeUser(),
});

const fakeCartItem = (overrides?: Object) => ({
  __typename: "CartItem",
  id: "omg123",
  quantity: 3,
  product: fakeItem(),
  user: fakeUser(),
  ...overrides,
});

function makePaginationMocksFor(length) {
  return [
    {
      request: { query: PAGINATION_QUERY },
      result: {
        data: {
          _allProductsMeta: {
            count: length,
          },
          itemsConnection: {
            __typename: "aggregate",
            aggregate: {
              count: length,
              __typename: "count",
            },
          },
        },
      },
    },
  ];
}

export {
  makePaginationMocksFor,
  fakeItem,
  fakeUser,
  fakeCartItem,
  fakeOrder,
  fakeOrderItem,
};
