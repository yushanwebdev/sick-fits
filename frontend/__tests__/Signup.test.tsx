import { MockedProvider } from "@apollo/client/testing";
import { render, screen } from "@testing-library/react";
import SignUp, { SIGN_UP_MUTATION } from "../components/SignUp";
import { fakeUser } from "../lib/testUtils";
import userEvent from "@testing-library/user-event";

const me = fakeUser();

const password = "wes";

const mocks = [
  // Mutation Mock
  {
    request: {
      query: SIGN_UP_MUTATION,
      variables: {
        name: me.name,
        email: me.email,
        password,
      },
    },
    result: {
      data: {
        createUser: {
          __typename: "User",
          id: "abc123",
          email: me.email,
          name: me.name,
        },
      },
    },
  },
];

describe("<SignUp/>", () => {
  it("render and matches snapshot", () => {
    const { container } = render(
      <MockedProvider>
        <SignUp />
      </MockedProvider>
    );

    expect(container).toMatchSnapshot();
  });

  it("calls the mutation properly", async () => {
    const { container, debug } = render(
      <MockedProvider mocks={mocks}>
        <SignUp />
      </MockedProvider>
    );

    // Type into the boxes
    await userEvent.type(screen.getByPlaceholderText("Your Name"), me.name);
    await userEvent.type(
      screen.getByPlaceholderText("Your Email Address"),
      me.email
    );
    await userEvent.type(screen.getByPlaceholderText("Password"), password);

    // Click the submit
    await userEvent.click(screen.getByText("Sign Up!"));
    await screen.findByText(
      `Signed Up with ${me.email} - Please Go Ahead and Sign in!`
    );
  });
});
