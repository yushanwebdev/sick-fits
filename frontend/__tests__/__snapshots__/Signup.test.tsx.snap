// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SignUp/> render and matches snapshot 1`] = `
<div>
  <form
    class="sc-23a63c62-0 bqKLMU"
    method="POST"
  >
    <h2>
      Sign Up For an Account
    </h2>
    <fieldset>
      <label
        for="name"
      >
        Name
        <input
          autocomplete="name"
          name="name"
          placeholder="Your Name"
          type="text"
          value=""
        />
      </label>
      <label
        for="email"
      >
        Email
        <input
          autocomplete="email"
          name="email"
          placeholder="Your Email Address"
          type="email"
          value=""
        />
      </label>
      <label
        for="password"
      >
        Password
        <input
          autocomplete="password"
          name="password"
          placeholder="Password"
          type="password"
          value=""
        />
      </label>
      <button
        type="submit"
      >
        Sign Up!
      </button>
    </fieldset>
  </form>
</div>
`;
